multiversx_sdk_cli/native_auth_client.py

```diff
@@ -0,0 +1,104 @@
+import base64
+import json
+from typing import Any, Dict, Optional
+
+import requests
+
+from multiversx_sdk_cli.errors import NativeAuthClientError
+
+EXPIRY_TIME_IN_SECONDS = 60 * 60 * 2
+DEFAULT_API_URL = "https://api.multiversx.com"
+
+
+class NativeAuthClientConfig:
+    def __init__(
+            self,
+            origin: str = '',
+            api_url: str = DEFAULT_API_URL,
+            expiry_seconds: int = EXPIRY_TIME_IN_SECONDS,
+            block_hash_shard: Optional[int] = None,
+            gateway_url: Optional[str] = None,
+            extra_request_headers: Optional[Dict[str, str]] = None
+    ) -> None:
+        self.origin = origin
+        self.api_url = api_url
+        self.expiry_seconds = expiry_seconds
+        self.block_hash_shard = block_hash_shard
+        self.gateway_url = gateway_url
+        self.extra_request_headers = extra_request_headers
+
+
+class NativeAuthClient:
+    def __init__(self, config: Optional[NativeAuthClientConfig] = None) -> None:
+        self.config = config or NativeAuthClientConfig()
+
+    def get_token(self, address: str, token: str, signature: str) -> str:
+        encoded_address = self.encode_value(address)
+        encoded_token = self.encode_value(token)
+
+        return f"{encoded_address}.{encoded_token}.{signature}"
+
+    def initialize(self, extra_info: Dict[Any, Any] = {}) -> str:
+        block_hash = self.get_current_block_hash()
+        encoded_extra_info = self.encode_value(json.dumps(extra_info))
+        encoded_origin = self.encode_value(self.config.origin)
+
+        return f"{encoded_origin}.{block_hash}.{self.config.expiry_seconds}.{encoded_extra_info}"
+
+    def get_current_block_hash(self) -> str:
+        if self.config.gateway_url:
+            return self._get_current_block_hash_using_gateway()
+        return self._get_current_block_hash_using_api()
+
+    def _get_current_block_hash_using_gateway(self) -> str:
+        round = self._get_current_round()
+        url = f"{self.config.gateway_url}/blocks/by-round/{round}"
+        response = self._execute_request(url)
+        blocks = response["data"]["blocks"]
+        block = [b for b in blocks if b["shard"] == self.config.block_hash_shard][0]
```

Comment:
A bit fragile - it's possible to have a round where not all shards propose a block (missed block).

Reply:
However, in the context of mxpy, we don't provide the shard.